shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -36,62 +36,72 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
-            S(mn -> MethodType(defns.head.level,
-              err(msg"A method definition must be given when inheriting multiple method definitions" -> tn.toLoc
-                :: msg"Definitions of method $mn inherited from:" -> N
-                :: defns.iterator.map(mt => msg"• ${mt.prts.head}" -> mt.prov.loco).toList),
-              tn :: Nil
-            ))
+            err(msg"An overriding method definition must be given when inheriting from multiple method definitions" -> tn.toLoc
+              :: msg"Definitions of method $mn inherited from:" -> N
+              :: defns.iterator.map(mt => msg"• ${mt.parents.head}" -> mt.prov.loco).toList)
+            S(mn -> MethodType(defns.head.level, N, tn :: Nil))
         }
       )
     private def ++(that: MethodDefs): MethodDefs = {
-      require(prts.isEmpty)
+      require(parents.isEmpty)
       MethodDefs(
         that.tn,
-        that.prts,
+        that.parents,
         this.decls ++ that.decls,
         this.defns ++ that.defns
       )
     }
-    private def addprt(prt: TypeName): MethodDefs =
-      copy(decls = decls.view.mapValues(mt => mt.copyMT(prts = prt :: mt.prts)).toMap)
+    private def addParent(prt: TypeName): MethodDefs =
+      copy(decls = decls.view.mapValues(mt => mt.copy(parents = prt :: mt.parents)).toMap)
     def propagate(top: Bool = true)(implicit raise: Raise): MethodDefs =
-      prts.map(_.propagate(false)).reduceOption(_.&(_)(tn, defns.keySet)).map(_.addprt(tn))
-        .foldRight(if (top) copy(prts = Nil, decls = Map.empty, defns = Map.empty) else copy(prts = Nil))(_ ++ _)
+      parents.map(_.propagate(false)).reduceOption(_.&(_)(tn, defns.keySet)).map(_.addParent(tn))
+        .foldRight(if (top) copy(parents = Nil, decls = Map.empty, defns = Map.empty) else copy(parents = Nil))(_ ++ _)
   }
   
-  case class Ctx(parent: Opt[Ctx], env: MutMap[Str, TypeScheme], mthenv: MutMap[Str, MethodType],
+  /** Keys of `mthEnv`:
+   *  `L` represents the inferred types of method definitions. The first value is the parent name, and the second value is the method name.
+   *  `R` represents the actual method types. The first optional value is the parent name, with `N` representing implicit calls,
+   *    and the second value is the method name.
+   *  The public helper functions should be preferred for manipulating `mthEnv`
+   */
+  case class Ctx(parent: Opt[Ctx], env: MutMap[Str, TypeInfo], mthEnv: MutMap[(Str, Str) \/ (Opt[Str], Str), MethodType],
       lvl: Int, inPattern: Bool, tyDefs: Map[Str, TypeDef]) {
-    def +=(b: Binding): Unit = env += b
-    def ++=(bs: IterableOnce[Binding]): Unit = bs.iterator.foreach(+=)
-    def get(name: Str): Opt[TypeScheme] = env.get(name) orElse parent.dlof(_.get(name))(N)
-    def getmth(name: Str): Opt[MethodType] = mthenv.get(name) orElse parent.dlof(_.getmth(name))(N)
+    def +=(b: Str -> TypeInfo): Unit = env += b
```

Comment:
Why expand the `Binding` alias everywhere?

Reply:
This was needed when the `TypeInfo` base class was introduced, and `env` was changed to map to `TypeInfo`s. We could either expand it here, or change the definition of the `Binding` alias and expand the reference in the return type of `typeStatement`, which only returns and is required to return a binding to a `TypeScheme`. I chose the former based on my interpretation of the word "binding", as a `TypeInfo` is not strictly a "binding". I can change it to the other way if you think that's more appropriate.

Reply:
Hmm okay, I guess that's fine.